### PR TITLE
 adsp: hda: Add support for buffers in virtual memory

### PR DESCRIPTION
**Pull request by softwarecki**
Removed the check whether a buffer is in the sram address space to allow use of buffers placed in the virtual memory address space.

### CODE REVIEW

**Review by andyross — COMMENTED** [review]
I don't see any virtualization going on here?  Surely these are **physical** addresses being passed, no?  And those need to be in appropriate physical regions for the hardware.  If you want to allow passing a virtual address you need to loop it through the MMU layer somehow to translate; needless to say the DSP doesn't have an IOMMU...

Maybe this is a homeless patch being submitted ahead of some other changes?  Or maybe "virtual" means something different in context?

**Comment by ceolin** [comment]
> I don't see any virtualization going on here? Surely these are **physical** addresses being passed, no? And those need to be in appropriate physical regions for the hardware. If you want to allow passing a virtual address you need to loop it through the MMU layer somehow to translate; needless to say the DSP doesn't have an IOMMU...
> 
> Maybe this is a homeless patch being submitted ahead of some other changes? Or maybe "virtual" means something different in context?

Yep, I don't know what **virtual memory** memory means in this context as well ..

**Comment by marcinszkudlinski** [comment]
@andyross @ceolin In case of MTL we don't have MMU unit in DSP CPU. All translations between physical and virtual address set is performed in TLB (Translation Lookup Buffer), a device located at front of memory banks. All requests - including from DMA and CPU - must go through TLB, there's no other way. 
So - unlike the whole world, in case of Intel Audio DSP DMAs use **virtual** address space, 1:1 matching CPUs

**Review by marcinszkudlinski — DISMISSED** [review]

**Review by kv2019i — DISMISSED** [review]

**Comment by andyross** [comment]
Is there going to be a driver for that TLB submitted at some point?  Maybe we can merge this then, conditional on support for it being enabled?  Again the assertions as written are correct as per the hardware zephyr understands.

**Comment by softwarecki** [comment]
> Is there going to be a driver for that TLB submitted at some point? 

@andyross The tlb driver exist for a long time and is still used. https://github.com/zephyrproject-rtos/zephyr/blob/f71a4184f3025ac77b6e17d0fa6d1386f20f23f3/drivers/mm/mm_drv_intel_adsp_mtl_tlb.c

**Comment by andyross** [comment]
Ah, cool.  OK then please predicate the asserts on the relevant kconfig(s) so that platforms that don't use it (looks like ace20_lnl is the only one that builds with it?) continue to check the addresses.

**Comment by andyross** [comment]
Also clarifying in the commit message that there is a non-userspace CPU mapping driver in use wouldn't hurt.  Almost everyone who sees the term "virtual memory" is going to assume you're talking about a userspace app.

**Comment by andyross** [comment]
Final point: **why** would you want to remap CPU access to DMA buffers?  Normally you want it the other way around, so a non-contiguous OS-managed set of pages can be mapped linearly for the DMA hardware to see.  I guess I don't understand the point of wanting to change the driver's (!!) view of the hardware (!!!) DMA memory layout, smells like a footgun.

**Comment by marcinszkudlinski** [comment]
>>Final point: why would you want to remap CPU access to DMA buffers? 

again, there's no other way to access the memory but using virtual adresses
In fact, there's nothing like "physical address" in CAVS/ACE architecture

The HW:

```
CPU ---|
       |---->(virtual addr space - 16MB) TLB ----> (mem bank number)Set_of_physical_memory_banks, less than 16MB
DMA ---|
```

what TLB does - it is just connecting a physical bank to a virtual address. In fact it operates on bank number, not a physical address. So - there will never ever be any access to "physical" because there's nothing like 

Tensilica core used in ACE doesn't have Tensilica MMU unit built-in, so there's no remapping at this level:

```
CPU (virtual) tensilica MMU ---|
                               |---->(physical)
DMA ---------------------------|
```

As for today, we use TLB very simply - just map all physical banks to start of virtual addr space:
```
virtual_space
 . (not_mapped)
 . (not_mapped)
 . (not_mapped)
 . (not_mapped)
 . (not_mapped)
 . ---- phys_bank_last
 (.....)
 . ---- phys_bank_3
 . ---- phys_bank_2
 . ---- phys_bank_1
```

so, the asserts in question are just covering the virtual space from lowest addr to the last **currently mapped**
but we need to start use it more widely, i.e. 
```
virtual_space
 . (not_mapped)
 . ---- phys_bank_7
 . (not_mapped)
 . ---- phys_bank_3
 . (not_mapped)
 . ---- phys_bank_10
 (.....)
 . ---- phys_bank_4
 . (not_mapped)
 . ---- phys_bank_1
```
to the space covered by asserts must be wider, and that's what the PR does

**Review by tmleman — DISMISSED** [review]

**Review by andyross — DISMISSED** [review]
OK, this is making more sense to me, and has the advantage of leaving validity checks in place.  Pedantic note: CONFIG_KERNEL_VM_* are set only when the KERNEL_VM subsystem is, which is selected by the MM_DRV kconfig, which is set only as a defconfig (i.e. it can be turned off). Now if you try that it will fail to build here.

**Review by pjdobrowolski — DISMISSED** [review]

**Review by gbernatxintel — DISMISSED** [review]

**Review by ceolin — DISMISSED** [review]

**Comment by marc-hb** [comment]
You just pushed a merge commit, probably not intended: https://github.com/zephyrproject-rtos/zephyr/pull/72511/commits/a75793491df11ea6ff24ef5c52acd445f6c7ea35

**Comment by softwarecki** [comment]
@lgirdwood can we merge it?

**Comment by kv2019i** [comment]
@softwarecki Merged for Zephyr are handled by Zephyr maintainers. Looping in @nashif 

I see at least one mandatory CI test still failing ("Run tests with twister") that probable needs to be addressed.

**Review by marc-hb — COMMENTED** [review]
Why the 100% copy/paste/diverge? How about making only the difference conditional?

```
#if defined(CONFIG_KERNEL_VM_SUPPORT)
#  define _INTEL_ADSP_BASE  CONFIG_KERNEL_VM_BASE
#  define _INTEL_ADSP_SIZE  CONFIG_KERNEL_VM_SIZE
#else
#  define _INTEL_ADSP_BASE  CONFIG_SRAM_BASE_ADDRESS
#  define _INTEL_ADSP_SIZE  CONFIG_SRAM_SIZE
#endif

  __ASSERT( ...
```

**Comment by softwarecki** [comment]
@marc-hb, @ceolin, @pjdobrowolski, @tmleman, @gbernatxintel, @andyross please review.

**Review by pjdobrowolski — APPROVED** [review]

**Review by marcinszkudlinski — APPROVED** [review]

**Review by dnikodem — APPROVED** [review]

**Review by serhiy-katsyuba-intel — APPROVED** [review]

**Review by tmleman — APPROVED** [review]

**Review by nashif — APPROVED** [review]

File: soc/intel/intel_adsp/common/include/intel_adsp_hda.h
Comment (marc-hb) [inline soc/intel/intel_adsp/common/include/intel_adsp_hda.h:175]: So this value is apparently 1024 times too small. 

- #74270

I really wonder how this commit was tested.

**Pull request merged**
